src/net/tcp/stream.rs

```diff
@@ -131,23 +145,30 @@ impl TcpStream {
     /// ```
     ///
     /// [`write`]: Self::write
-    pub async fn write_all<T: IoBuf>(&self, mut buf: T) -> crate::BufResult<(), T> {
-        let mut n = 0;
-        while n < buf.bytes_init() {
-            let res = self.write(buf.slice(n..)).await;
+    pub async fn write_all<T: IoBuf>(
+        &self,
+        buf: impl Into<Slice<T>>,
+    ) -> crate::BufResult<(), Slice<T>> {
+        self.write_all_slice(buf.into()).await
+    }
+
+    // The implementation of write_all, factored out to reduce code bloat.
+    async fn write_all_slice<T: IoBuf>(&self, mut buf: Slice<T>) -> crate::BufResult<(), Slice<T>> {
```

Comment:
Nice catch. @Noah-Kennedy has said methods like write_all will be moved to a trait some time in the future.